runtime/tests/checker/transactions_test.go

```diff
@@ -398,82 +727,171 @@ func TestCheckInvalidTransactionSelfMoveToFunction(t *testing.T) {
-	assert.IsType(t, &sema.TransactionType{}, typeMismatchErr.ActualType)
-}
+		require.IsType(t, &sema.TypeMismatchError{}, errs[0])
+		typeMismatchErr := errs[0].(*sema.TypeMismatchError)
 
-func TestCheckInvalidTransactionSelfMoveIntoArrayLiteral(t *testing.T) {
+		assert.Equal(t, sema.VoidType, typeMismatchErr.ExpectedType)
+		assert.IsType(t, &sema.TransactionType{}, typeMismatchErr.ActualType)
+	})
 
-	t.Parallel()
+	t.Run("into array literal", func(t *testing.T) {
 
-	_, err := ParseAndCheck(t, `
+		t.Parallel()
 
-     transaction {
+		_, err := ParseAndCheck(t, `
 
-         execute {
-             let txs = [self]
-         }
-     }
-   `)
+          transaction {
 
-	errs := RequireCheckerErrors(t, err, 1)
+              execute {
+                  let txs = [self]
+              }
+          }
+        `)
 
-	assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		errs := RequireCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+	})
+
+	t.Run("into dictionary literal", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+
+          transaction {
+
+              execute {
+                  let txs = {"self": self}
+              }
+          }
+        `)
+
+		errs := RequireCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+	})
 }
 
-func TestCheckInvalidTransactionSelfMoveIntoDictionaryLiteral(t *testing.T) {
+func TestCheckInvalidTransactionRoleSelfMove(t *testing.T) {
 
 	t.Parallel()
 
-	_, err := ParseAndCheck(t, `
+	t.Run("variable declaration", func(t *testing.T) {
 
-     transaction {
+		_, err := ParseAndCheck(t, `
 
-         execute {
-             let txs = {"self": self}
-         }
-     }
-   `)
+          transaction {
 
-	errs := RequireCheckerErrors(t, err, 1)
+              role buyer {
+                  prepare() {
+                      let x = self
+                  }
+              }
+          }
+        `)
 
-	assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+		errs := RequireCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+	})
+
+	t.Run("into array literal", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+
+          transaction {
+
+              role buyer {
+                  prepare() {
+                      let txs = [self]
+                  }
+              }
+          }
+        `)
+
+		errs := RequireCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+	})
+
+	t.Run("into dictionary literal", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+
+          transaction {
+
+              role buyer {
+                  prepare() {
+                      let txs = {"self": self}
+                  }
+              }
+          }
+        `)
+
+		errs := RequireCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidMoveError{}, errs[0])
+	})
+
+	t.Run("role in execute", func(t *testing.T) {
```

Comment:
The `checkValueMove` is only get called only for a limited set of expressions. Maybe add some tests for other use-cases, such as calling a function on a role, etc.

Reply:
Improved the tests in 36f2848657ff94ea58f1770d221a85232b66fb31